### PR TITLE
feat: Modules v2 support

### PR DESCRIPTION
**Pull request by disq**

### CODE REVIEW

File: pkg/module/manager.go [hunk 1]
```diff
@@ -81,3 +99,93 @@ func (m *ManagerImpl) ExampleConfigs() []string {
...
+	if !foundUnsupported {
+		// happy path
+		return doVersions[0], ret, nil
+	}
+
+	if forceVersion > 0 {
+		return 0, nil, errNegotiationFailed
+	}
+
+	// negotiate: through each version supported by the module, in order of prefence, and try to find an entry which all providers can satisfy
```
Comment (roneli) [inline pkg/module/manager.go:150]:
```suggestion
	// negotiate: through each version supported by the module, in order of preference, and try to find an entry which all providers can satisfy
```

File: pkg/module/manager.go [hunk 2]
```diff
@@ -60,7 +69,16 @@ func (m *ManagerImpl) ExecuteModule(ctx context.Context, modName string, cfg hcl
 		return nil, fmt.Errorf("module not found %q", modName)
 	}
 
-	if err := mod.Configure(ctx, cfg, execReq.Params); err != nil {
```
Comment (roneli) [inline pkg/module/manager.go:63]: ill comment here as I can't comment on the ExecuteModule function, why we pass the name and config seperatly and then put providers and params in the request (i.e pass th config+ name there)

File: pkg/module/manager.go [hunk 3]
```diff
@@ -60,7 +69,16 @@ func (m *ManagerImpl) ExecuteModule(ctx context.Context, modName string, cfg hcl
...
+	protoVersion, modInfo, err := m.negotiateProtocolVersions(ctx, mod, execReq.Providers, 0)
+	if err != nil {
+		return nil, fmt.Errorf("protocol negotiation failed: %w", err)
+	}
+
+	if err := mod.Configure(ctx, Info{
```
Comment (roneli) [inline pkg/module/manager.go:77]: this doesn't seem thread safe TBH, every execution should create a new module instance.

File: pkg/module/manager.go [hunk 4]
```diff
@@ -81,3 +99,93 @@ func (m *ManagerImpl) ExampleConfigs() []string {
...
+	}
+
+	var (
+		ret               = make(map[string]ProviderData, len(provs)) // provider vs. info-key vs. files provided by provider under that key
+		supportedVersions = make(map[string][]uint32, len(provs))
+		foundUnsupported  bool
+	)
+
+	// Do initial requests
+	for _, p := range provs {
```
Comment (roneli) [inline pkg/module/manager.go:120]: I think the negotiation should be in the rpc level and not need for two loops, i.e give me versions X, and you get them back. Moreover we can do that request in the client and pass the data in the execRequest (the module info). If we don't support any versions prior to the call we shouldn't call the execute. (It couples the client with the modules a little, but the issue here is the general design of the client, and I don't think we should go deep and refactor that now).

**Review by roneli — CHANGES_REQUESTED** [review]
Comments on modules v2 support

File: pkg/module/manager.go [hunk 5]
```diff
@@ -60,7 +69,16 @@ func (m *ManagerImpl) ExecuteModule(ctx context.Context, modName string, cfg hcl
 		return nil, fmt.Errorf("module not found %q", modName)
 	}
 
-	if err := mod.Configure(ctx, cfg, execReq.Params); err != nil {
+	protoVersion, modInfo, err := m.negotiateProtocolVersions(ctx, mod, execReq.Providers, 0)
```
[the remark this answers is not in the document]
Comment (disq) [reply pkg/module/manager.go:72]: it's used in negotiating.

File: pkg/module/drift/config_parser.go [hunk 6]
```diff
@@ -86,7 +86,7 @@ func replacePlaceholderInSlice(varName placeholder, value, subject []string) []s
 	return newSubj
 }
 
-func (p *Parser) Decode(body hcl.Body, diags hcl.Diagnostics) (*BaseConfig, hcl.Diagnostics) {
+func (p *Parser) Decode(body hcl.Body, allowedProvider *string, diags hcl.Diagnostics) (*BaseConfig, hcl.Diagnostics) {
```
[the remark this answers is not in the document]
Comment (disq) [reply pkg/module/drift/config_parser.go:89]: sure

File: pkg/module/drift/config_parser.go [hunk 7]
```diff
@@ -110,6 +110,16 @@ func (p *Parser) Decode(body hcl.Body, diags hcl.Diagnostics) (*BaseConfig, hcl.
 			prov, provDiags := p.decodeProviderBlock(block, &ctx)
 			diags = append(diags, provDiags...)
 			if prov != nil {
+				if allowedProvider != nil && prov.Name != *allowedProvider {
```
[the remark this answers is not in the document]
Comment (disq) [reply pkg/module/drift/config_parser.go:113]: if inverted, let's not change the way it works any more for now

File: pkg/module/types.go [hunk 8]
```diff
@@ -6,19 +6,36 @@ import (
...
 	Execute(context.Context, *ExecuteRequest) *ExecutionResult
 	// ExampleConfig returns an example configuration to be put in config.hcl
 	ExampleConfig() string
 }
 
+// Info about user supplied configs and provider supplied module info
+type Info struct {
+	// UserConfig is the config supplied by the user in config.hcl
+	UserConfig hcl.Body
+
```
[the remark this answers is not in the document]
Comment (disq) [reply pkg/module/types.go:29]: not rogue, intended.

File: pkg/module/manager.go [hunk 9]
```diff
@@ -81,3 +99,93 @@ func (m *ManagerImpl) ExampleConfigs() []string {
 	}
 	return ret
 }
+
+var errNegotiationFailed = fmt.Errorf("version mismatch between module and providers, please upgrade your provider and/or cloudquery")
```
[the remark this answers is not in the document]
Comment (disq) [reply pkg/module/manager.go:103]: fixed

File: pkg/module/manager.go [hunk 10]
```diff
@@ -34,12 +38,17 @@ type Manager interface {
 	ExampleConfigs() []string
 }
 
+type moduleInfoQuerier interface {
```
[the remark this answers is not in the document]
Comment (disq) [reply pkg/module/manager.go:41]: fixed

File: pkg/module/manager.go (hunk 2 above)
Comment (disq) [reply pkg/module/manager.go:63]: fixed

**Review by roneli — APPROVED** [review]
LGTM

File: pkg/module/drift/drift_test.go [hunk 11]
```diff
@@ -5,23 +5,76 @@ import (
...
+	"github.com/cloudquery/cloudquery/pkg/module"
+	"github.com/cloudquery/cq-provider-sdk/cqproto"
 	"github.com/doug-martin/goqu/v9"
 	"github.com/doug-martin/goqu/v9/exp"
 	"github.com/hashicorp/go-hclog"
 	"github.com/hashicorp/hcl/v2/hclparse"
 	"github.com/stretchr/testify/assert"
 )
 
+func setupDrift(t *testing.T) (*Drift, *BaseConfig) {
```
[the remark this answers is not in the document]
Comment (disq) [reply pkg/module/drift/drift_test.go:16]: added

File: pkg/module/manager.go [hunk 12]
```diff
@@ -142,5 +143,61 @@ func (m *ManagerImpl) collectProviderInfo(ctx context.Context, mod Module, provs
...
+		return 0, nil, fmt.Errorf("providers %s seem to support an older version of %s, which is incompatible with your cloudquery version", strings.Join(olderProviders, ", "), mod.ID())
+	}
+
+	if l := len(newerProviders); l == 1 {
+		return 0, nil, fmt.Errorf("provider %s seems to support a newer version of %s, which is incompatible with your cloudquery version", newerProviders[0], mod.ID())
+	} else if l > 1 {
+		return 0, nil, fmt.Errorf("providers %s seem to support a newer version of %s, which is incompatible with your cloudquery version", strings.Join(newerProviders, ", "), mod.ID())
+	}
+
+	return 0, nil, fmt.Errorf("version mismatch between module and providers, please upgrade your providers and/or cloudquery")
```
Comment (roneli) [inline pkg/module/manager.go:182]: nit: put this in a function?

Comment (disq) [reply pkg/module/manager.go]: done, simplified and added test

**Review by roneli — APPROVED** [review]

**Pull request merged**